src/collapse/dtrace.rs

```diff
@@ -113,19 +120,129 @@ impl Collapse for Folder {
+
+        // skip header lines -- first empty line marks start of data
+        loop {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                // We reached the end :( this should not happen.
+                warn!("File ended while skipping headers");
+                return Ok(());
+            };
+            if line.trim().is_empty() {
+                break;
+            }
+        }
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        debug_assert!(self.occurrences.is_concurrent());
+        debug_assert!(self.opt.nthreads > 1);
+
+        let mut buf = Vec::with_capacity(CAPACITY_INPUT_BUFFER);
+        reader.read_to_end(&mut buf)?;
+
+        let mut input = Input::new(buf, self.opt.nthreads, Self::identify_stack_locations)?;
+
+        crossbeam::thread::scope(|scope| {
+            let mut handles = Vec::with_capacity(input.nthreads());
+            let (sender, receiver) = crossbeam::channel::bounded(input.nthreads());
+            for chunk in input.chunks() {
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt;
+                let sender = sender.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        // state
+                        cache_inlines: Vec::new(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+
+                        // options
+                        opt,
+                    };
+                    let result = folder.collapse_single_threaded(chunk);
+                    sender.send(result).unwrap();
+                });
+                handles.push(handle);
+            }
+            for handle in handles {
+                receiver.recv().unwrap()?;
+                handle.join().unwrap();
+            }
+            Ok::<_, io::Error>(())
+        })
+        .unwrap()?;
+
+        Ok(())
+    }
+
+    // When collapsing using multiple threads, this function is run once
+    // upfront in order to determine the locations (byte indices) of each
+    // stack in the input data so that the input data can later be broken up
+    // into pieces in order to be allocated across multiple threads.
+    // See `crate::collapse::Input::new`.
+    fn identify_stack_locations(mut reader: io::BufReader<&[u8]>) -> io::Result<Vec<usize>> {
```

Comment:
I wonder if there's a way for us to share this among multiple collapsers by having them implement some kind of `identify_boundary` method. Then we could use the same `Input` impl across all of them.